controller/runner.go

```diff
@@ -0,0 +1,186 @@
+	mu   sync.Mutex
+
+	server     *grpc.Server
+	shipper    server.ShipperServer
+	queue      *queue.Queue
+	monitoring *monitoring.QueueMonitor
+	out        Output
+}
+
+// NewServerRunner creates a new runner that starts and stops the server.
+// Consumers are expected to run `Close` to release all resources created
+// by a successful result from this function even without starting the server.
+func NewServerRunner(cfg config.ShipperConfig) (r *ServerRunner, err error) {
+	r = &ServerRunner{
+		log: logp.L(),
+		cfg: cfg,
+	}
+	// in case of an initialization error we must clean up all created resources
+	defer func() {
+		if err != nil {
+			// this will account for partial initialization in case of an error, so there are no leaks
+			r.Close()
+		}
+	}()
+
+	r.log.Debug("initializing the queue...")
+	r.queue, err = queue.New(cfg.Queue)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create queue: %w", err)
+	}
+	r.log.Debug("queue was initialized.")
+
+	r.log.Debug("initializing monitoring ...")
+	r.monitoring, err = monitoring.NewFromConfig(cfg.Monitor, r.queue)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing output monitor: %w", err)
+	}
+	r.monitoring.Watch()
+	r.log.Debug("monitoring is ready.")
+
+	r.log.Debug("initializing the output...")
+	// TODO replace with the real output based on the config, Console is hard-coded for now
+	r.out = output.NewConsole(r.queue)
+	r.out.Start()
+	r.log.Debug("output was initialized.")
+
+	r.log.Debug("initializing the gRPC server...")
+	var opts []grpc.ServerOption
+	if cfg.Server.TLS {
+		creds, err := credentials.NewServerTLSFromFile(cfg.Server.Cert, cfg.Server.Key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate credentials %w", err)
+		}
+		opts = []grpc.ServerOption{grpc.Creds(creds)}
+	}
+	r.server = grpc.NewServer(opts...)
+	r.shipper, err = server.NewShipperServer(cfg.Server, r.queue)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialise the server: %w", err)
+	}
+	pb.RegisterProducerServer(r.server, r.shipper)
+	r.log.Debug("gRPC server initialized.")
+
+	return r, nil
+}
+
+// Start runs the shipper server according to the set configuration.
+// The server stops after calling `Close`, this function blocks until then.
+func (r *ServerRunner) Start() (err error) {
+	r.mu.Lock()
+	if r.server == nil {
+		r.mu.Unlock()
+		return fmt.Errorf("failed to start a server runner that was previously closed")
+	}
+
+	addr := fmt.Sprintf("localhost:%d", r.cfg.Server.Port)
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		r.mu.Unlock()
+		return fmt.Errorf("failed to listen on %s: %w", addr, err)
+	}
+
+	// Testing that the server is running and only then unlock the mutex.
+	// Otherwise if `Close` is called at the same time as `Start` it causes race condition.
+	go func() {
+		defer r.mu.Unlock()
+		con, err := net.Dial("tcp", addr)
+		if err != nil {
+			r.log.Errorf("failed to test connection with the gRPC server on %s", addr)
+			return
```

Comment:
Do we really want to keep running if we can't connect to the server?  If we take this out of the go func, we could return the error.

Reply:
I don't see how we can report an error from this go routine other than just logging it. This function returns on running the server (blocking) that this go routine is checking connection to.

Reply:
It probably doesn't make sense to put an essential connection test somewhere that the code can't react to the result. Between this difficulty and the concurrency conflict between Start / Close, maybe an alternate pattern would be better:

If`Start` _doesn't_ block on the server (but instead calls `go r.server.Serve(lis)`) then we can instead have it block on the connection test, so we can propagate the error upward if the test fails.

This would also give us the chance to relax our concurrency requirements -- if `r.Start` is expected to be called synchronously, then we can probably skip the mutex / once protections that keep it from interacting with `Close`. The default expectation for a Go object is that it shouldn't be shared between goroutines without synchronization, and as far as I know we don't need anything stronger than that, so we can assume that all of the methods on `ServerRunner` will be called synchronously if it simplifies the design.